// @redocly/openapi-core/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bundle = exports.lintDocument = exports.validate = exports.lint = exports.OasVersion = exports.formatProblems = exports.walkDocument = exports.normalizeVisitors = exports.openAPIMajor = exports.OasMajorVersion = exports.detectOpenAPI = exports.unescapePointer = exports.YamlParseError = exports.ResolveError = exports.resolveDocument = exports.BaseResolver = exports.RedoclyClient = exports.LintConfig = exports.Config = exports.loadConfig = exports.Stats = exports.normalizeTypes = exports.Oas2Types = exports.Oas3Types = void 0;
var oas3_1 = require("./types/oas3");
Object.defineProperty(exports, "Oas3Types", { enumerable: true, get: function () { return oas3_1.Oas3Types; } });
var oas2_1 = require("./types/oas2");
Object.defineProperty(exports, "Oas2Types", { enumerable: true, get: function () { return oas2_1.Oas2Types; } });
var types_1 = require("./types");
Object.defineProperty(exports, "normalizeTypes", { enumerable: true, get: function () { return types_1.normalizeTypes; } });
var stats_1 = require("./rules/other/stats");
Object.defineProperty(exports, "Stats", { enumerable: true, get: function () { return stats_1.Stats; } });
var config_1 = require("./config/config");
Object.defineProperty(exports, "loadConfig", { enumerable: true, get: function () { return config_1.loadConfig; } });
Object.defineProperty(exports, "Config", { enumerable: true, get: function () { return config_1.Config; } });
Object.defineProperty(exports, "LintConfig", { enumerable: true, get: function () { return config_1.LintConfig; } });
var redocly_1 = require("./redocly");
Object.defineProperty(exports, "RedoclyClient", { enumerable: true, get: function () { return redocly_1.RedoclyClient; } });
var resolve_1 = require("./resolve");
Object.defineProperty(exports, "BaseResolver", { enumerable: true, get: function () { return resolve_1.BaseResolver; } });
Object.defineProperty(exports, "resolveDocument", { enumerable: true, get: function () { return resolve_1.resolveDocument; } });
Object.defineProperty(exports, "ResolveError", { enumerable: true, get: function () { return resolve_1.ResolveError; } });
Object.defineProperty(exports, "YamlParseError", { enumerable: true, get: function () { return resolve_1.YamlParseError; } });
var ref_utils_1 = require("./ref-utils");
Object.defineProperty(exports, "unescapePointer", { enumerable: true, get: function () { return ref_utils_1.unescapePointer; } });
var lint_1 = require("./lint");
Object.defineProperty(exports, "detectOpenAPI", { enumerable: true, get: function () { return lint_1.detectOpenAPI; } });
Object.defineProperty(exports, "OasMajorVersion", { enumerable: true, get: function () { return lint_1.OasMajorVersion; } });
Object.defineProperty(exports, "openAPIMajor", { enumerable: true, get: function () { return lint_1.openAPIMajor; } });
var visitors_1 = require("./visitors");
Object.defineProperty(exports, "normalizeVisitors", { enumerable: true, get: function () { return visitors_1.normalizeVisitors; } });
var walk_1 = require("./walk");
Object.defineProperty(exports, "walkDocument", { enumerable: true, get: function () { return walk_1.walkDocument; } });
var format_1 = require("./format/format");
Object.defineProperty(exports, "formatProblems", { enumerable: true, get: function () { return format_1.formatProblems; } });
var lint_2 = require("./lint");
Object.defineProperty(exports, "OasVersion", { enumerable: true, get: function () { return lint_2.OasVersion; } });
Object.defineProperty(exports, "lint", { enumerable: true, get: function () { return lint_2.lint; } });
Object.defineProperty(exports, "validate", { enumerable: true, get: function () { return lint_2.lint; } });
Object.defineProperty(exports, "lintDocument", { enumerable: true, get: function () { return lint_2.lintDocument; } });
var bundle_1 = require("./bundle");
Object.defineProperty(exports, "bundle", { enumerable: true, get: function () { return bundle_1.bundle; } });
